src/consensus/parlia/mod.rs

```diff
@@ -0,0 +1,556 @@
+
+        let proposer = util::recover_creator(header, self.chain_id)?;
+        if proposer != header.beneficiary {
+            return Err(ParliaError::WrongHeaderSigner {
+                number: header.number,
+                expected: header.beneficiary,
+                got: proposer,
+            }.into());
+        }
+        if !snap.validators.contains(&proposer) {
+            return Err(ParliaError::SignerUnauthorized {
+                number: header.number,
+                signer: proposer,
+            }.into());
+        }
+        for (seen, recent) in snap.recent_proposers.iter() {
+            if *recent == proposer {
+                // Signer is among recent_proposers, only fail if the current block doesn't shift it out
+                let limit = (snap.validators.len() / 2 + 1) as u64;
+                if *seen > block_number.0 - limit {
+                    return Err(ParliaError::SignerOverLimit { signer: proposer }.into());
+                }
+            }
+        }
+        let inturn_proposer = snap.inturn(&proposer);
+        if inturn_proposer && header.difficulty != DIFF_INTURN {
+            return Err(ValidationError::WrongDifficulty.into());
+        } else if !inturn_proposer && header.difficulty != DIFF_NOTURN {
+            return Err(ValidationError::WrongDifficulty.into());
+        }
+        Ok(())
+    }
+
+    /// parlia's finalize not effect any state, must set transaction and ConsensusFinalizeState in sync
+    fn finalize(
+        &self,
+        header: &BlockHeader,
+        _ommers: &[BlockHeader],
+        transactions: Option<&Vec<MessageWithSender>>,
+        state: &dyn StateReader,
+    ) -> anyhow::Result<Vec<FinalizationChange>> {
+
+        // check epoch validators chg correctly
+        if self.new_block_state.parsed_validators() && header.number % self.epoch == 0 {
+            let expect_validators = self.new_block_state.get_validators()
+                .ok_or_else(|| ParliaError::CacheValidatorsUnknown)?;
+
+            let actual_validators = util::parse_epoch_validators(&header.extra_data[VANITY_LENGTH..(header.extra_data.len() - SIGNATURE_LENGTH)])?;
+
+            info!("epoch validators check {}, {}:{}", header.number, actual_validators.len(), expect_validators.len());
+            if actual_validators != *expect_validators {
+                return Err(ParliaError::EpochChgWrongValidators {
+                    expect: expect_validators.clone(),
+                    got: actual_validators,
+                }.into());
+            }
+        }
+
+        // if set transactions, check systemTxs and reward if correct
+        // must set transactions in sync
+        if let Some(transactions) = transactions {
+            let mut system_txs: Vec<&MessageWithSender> = transactions.iter()
+                .filter(|tx| is_system_transaction(&tx.message, &tx.sender, &header.beneficiary))
+                .collect();
+            if header.number == 1 {
```

Comment:
May not need to treat header 1 specially?

Reply:
got it, it is used to skip the 1st block, since there are no mined gas fee in this block.